### ISSUE TITLE
Metrics gattering patterns on plugins

### ISSUE THREAD

**Issue by purbon** [elastic/logstash#5716]
Hi,
  during the work I've been doing lately on #5607 I've noticed a few patterns that might DRY'ed somehow in order to provide:

* Usage of similar vocabulary in terms of, singulars vs plurals.
* Same words for common terms, for example to name errors, etc.
* Provide uniform structure, so consumers will expect similar organization.

I'm sure as far as we get into adding more metrics to filters and outputs we will encounter patterns that might makes sense to uniform.

### Patterns

1.  Dealing with errors happening during execution path. 

For this I started creating a namespaced metric, ``` @metric_errors = metric.namespace(:errors)``` that later on is used to track error situations like exceptions, for example. This makes all error situation be namespaced.

example:

<img width="507" alt="aa6c8e48-4e9a-11e6-80a6-ab60a918403e" src="https://cloud.githubusercontent.com/assets/68540/17298045/ac27b902-5808-11e6-8b86-2bae5c78b4d2.png">

**Comment by jordansissel** [comment]
Style guide proposal:

* lowercase
* try to have a single word be the name
* if two words are best, try to make them `name_unit`  where `name` describes the metric and `unit` describes the units the value is using.
* plural for counts of things: Using "event" as an example, where we would describe 'count of Event' we should use the plural `events`
* use namespaces instead of multiple words. For example, instead of `connection_errors` we can create an `errors` namespace with `connections` in it.
* namespace names should be plural and be the name of what they contain. For example, an `errors` namespace contains metrics for errors.
* namespaces hierarchy should put aspects/categories first, where possible. For an example: `{ "errors": { "connections": 5 }` is better than `{ "connections": { "errors": 5 } }`

**Comment by ph** [comment]
@jordansissel excellent writes up, I will finish the review of the PR and update this issue also.

I think we should name this as `best practices`, even if some of theses check could be done in the code. I think we shouldn't do it, since we want metric to be fast.

**Comment by jordansissel** [comment]
agreed, my though my preference is to call it a style guide, not best practices

**Comment by jordansissel** [comment]
What are others' thoughts on this style guide I proposed a few comments above?

**Comment by suyograo** [comment]
+1. Looks good

**Comment by purbon** [comment]
@jordansissel thanks for your feedback, will update the current PR's following your advices. I'm not sure about the naming you propose,  but I guess I'm good with it, basic idea for me was to extrapolate similar patterns for all to follow, so I guess naming is ok 😸 .

@ph what do you mean by

> even if some of theses check could be done in the code 

? curious to know more what you are thinking here.

**Comment by ph** [comment]
I've reread the plugins PR and I am +1 with your proposal @jordansissel.

> even if some of theses check could be done in the code

We could at the metric library normalize keys and adding validation.

**Comment by acchen97** [comment]
Thanks for this @jordansissel.

This style guide looks like good material to include within a general "Adding and Collecting New Plugin Metrics" section in the docs.

**Comment by jordansissel** [comment]
@acchen97 ++. Target this for 5.0.0?

**Comment by acchen97** [comment]
++ added to 5.0 docs meta issue to track.
